File: Data Structure/Stack/Hard/LET_no224_Basic Calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        res, sign, num = 0, 1, 0
        for c in s:
            if c.isdigit():
                num = 10 * num + int(c)
            elif c == '+':
                res += sign * num
                sign = 1
                num = 0
            elif c == '-':
                res += sign * num
                sign = -1
                num = 0
            elif c == '(':
                stack.append((res, sign))
                sign = 1
                res = 0
            elif c == ')':
                res += sign * num
                num = 0
                prev_res, prev_sign = stack.pop()
                res *= prev_sign
                res += prev_res
        
        if num > 0:
            res += sign * num
        
        return res

File: Data Structure/Stack/Hard/LET_no224_Basic Calculator.cpp
class Solution {
public:
    int calculate(string s) {
        vector<pair<int, int>> stack;
        int res = 0, sign = 1, num = 0;
        for (char c : s) {
            if (c >= '0')
                num = num * 10 + (c - '0');
            else if (c == '+' or c == '-') {
                res += sign * num;
                sign = c == '+' ? 1 : -1;
                num = 0;
            }
            else if (c == '(') {
                stack.push_back({res, sign});
                res = 0;
                sign = 1;
            }
            else if (c == ')') {
                res += sign * num;
                num = 0;
                pair<int, int> prev = stack.back();
                res *= prev.second;
                res += prev.first;
                stack.pop_back();
            }
        }
        
        if (num > 0)
            res += sign * num;
        
        return res;
    }
};